File: javascript/2015/day08a/day08a.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const process = require('process');

function usage() {
    const progname = path.basename(process.argv[1]);
    console.log(`usage: ${progname} <input file>`);
    process.exit(1);
}

function processFile(filename) {
    let result = 0;
    const data = fs.readFileSync(filename, 'utf8');
    const lines = data.trim().split('\n');
    for ( const line of lines ) {
        const codeLen = line.length;
        let memLen = 0;
        let i = 1;
        while ( i < line.length - 1 ) {
            switch ( line[i] ) {
                case '\\':
                    switch ( line[i + 1] ) {
                        case '\\':
                        case '"':
                            i += 2;
                            break;
                        case 'x':
                            i += 4;
                            break;
                        default:
                            i++;
                    }
                    break;
                default:
                    i++;
            }
            memLen++;
        }
        result += codeLen - memLen;
    }
    return result;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    const filename = process.argv[2];
    const result = processFile(filename);
    console.log(`result = ${result}`);
}

main()

File: ruby/2015/day08a/day08a.rb
#!/usr/bin/env ruby

def usage
    progname = File.basename($0)
    puts "usage: #{progname} <input file>"
    exit 1
end

def process(filename)
    result = 0
    File.foreach(filename) do |line|
        code_len = line.length
        mem_len = 0
        i = 1
        while i < line.length - 1 do
            case line[i]
            when "\\"
                case line[i + 1]
                when "\\", "\""
                    i += 2
                when "x"
                    i += 4
                else
                    i += 1
                end
            else
                i += 1
            end
            mem_len += 1
        end
        result += code_len - mem_len
    end
    result
end

def main
    usage if ARGV.length < 1
    filename = ARGV[0]
    result = process(filename)
    puts "result = #{result}"
end

main if __FILE__ == $0
